parsing/parser.mly

```diff
@@ -211,7 +211,7 @@ let mkexp_opt_constraint ~loc e = function
 
 let mkpat_opt_constraint ~loc p = function
   | None -> p
-  | Some typ -> mkpat ~loc (Ppat_constraint(p, typ))
+  | Some typ -> ghpat ~loc (Ppat_constraint(p, typ))
```

Comment:
Here for example, how does making this specific pattern ghost help with the Refis laws (I guess the second law)?

Reply:
This one helps with cases like `function { contents : int } -> ...` and is a bit tricky to understand by looking at the diff in the testsuite, because the printer pretends the pattern is a child of the label, which it is not:
```ocaml
| Ppat_record of (Longident.t loc * pattern) list * closed_flag
```
in this particular example, the pattern's and longident's locations overlap.
I made the arbitrary choice of deciding that in presence of punning, the pattern / expression isn't really there in the source, only the label is. That's why I chose to mark that node as ghost; but another acceptable choice would be to mark the longident as such (but I like it slightly less).